datafusion/core/tests/execution/coop.rs

```diff
@@ -0,0 +1,722 @@
+        )],
+        None,
+        &JoinType::Left,
+        None,
+        PartitionMode::CollectLeft,
+        true,
+    )?);
+
+    query_yields(join, session_ctx.task_ctx()).await
+}
+
+#[rstest]
+#[tokio::test]
+async fn hash_join_without_repartition_and_no_agg(
+    #[values(false, true)] pretend_infinite: bool,
+) -> Result<(), Box<dyn Error>> {
+    // Create Session, schema, and an 8K-row RecordBatch for each side
+    let session_ctx = SessionContext::new();
+
+    // on the right side, we’ll shift each value by +1 so that not everything joins,
+    // but plenty of matching keys exist (e.g. 0 on left matches 1 on right, etc.)
+    let infinite_left = make_lazy_exec_with_range("value", -10..10, false);
+    let infinite_right =
+        make_lazy_exec_with_range("value", 0..i64::MAX, pretend_infinite);
+
+    // Directly feed `infinite_left` and `infinite_right` into HashJoinExec.
+    // Do not use aggregation or repartition.
+    let join = Arc::new(HashJoinExec::try_new(
+        Arc::new(infinite_left),
+        Arc::new(infinite_right),
+        vec![(
+            Arc::new(Column::new("value", 0)),
+            Arc::new(Column::new("value", 0)),
+        )],
+        /* filter */ None,
+        &JoinType::Inner,
+        /* output64 */ None,
+        // Using CollectLeft is fine—just avoid RepartitionExec’s partitioned channels.
+        PartitionMode::CollectLeft,
+        /* build_left */ true,
+    )?);
+
+    query_yields(join, session_ctx.task_ctx()).await
+}
+
+async fn query_yields(
+    plan: Arc<dyn ExecutionPlan>,
+    task_ctx: Arc<TaskContext>,
+) -> Result<(), Box<dyn Error>> {
+    // Run plan through EnsureCooperative
+    let optimized =
+        EnsureCooperative::new().optimize(plan, task_ctx.session_config().options())?;
+
+    // Get the stream
+    let mut stream = physical_plan::execute_stream(optimized, task_ctx)?;
+
+    // Create an independent executor pool
+    let child_runtime = Runtime::new()?;
+
+    // Spawn a task that tries to poll the stream
+    // The task returns Ready when the stream yielded with either Ready or Pending
+    let join_handle = child_runtime.spawn(std::future::poll_fn(move |cx| {
+        match stream.poll_next_unpin(cx) {
+            Poll::Ready(_) => Poll::Ready(Poll::Ready(())),
+            Poll::Pending => Poll::Ready(Poll::Pending),
+        }
+    }));
+
+    let abort_handle = join_handle.abort_handle();
+
+    // Now select on the join handle of the task running in the child executor with a timeout
+    let yielded = select! {
+        _ = join_handle => true,
+        _ = tokio::time::sleep(tokio::time::Duration::from_secs(10)) => false
+    };
+
+    // Try to abort the poll task and shutdown the child runtime
+    abort_handle.abort();
+    Handle::current().spawn_blocking(move || {
+        drop(child_runtime);
+    });
+
+    // Finally, check if poll_next yielded
+    assert!(yielded, "Task did not yield in a timely fashion");
+    Ok(())
+}
```

Comment:
We also need to add spill merge case since we add the yield.

Reply:
Good point. Do you know a practical way we could set this up in a test? In order to observe the problem you would probably need to spill a substantial amount of data.

Reply:
maybe @2010YOUY01 or @ding-young have some ideas